File: Packages/BudgetKit/Sources/BudgetKit/KeypadEngine.swift
import Foundation

/// The keys on Budget Bestie's budgeting keypad.
public enum KeypadKey: Hashable, Sendable {
    case digit(Int)
    case doubleZero
    /// Backspace — removes the last typed digit, or clears the amount if none
    /// have been typed yet.
    case delete
    case add
    case subtract
    case equals
    /// Fills in the engine-calculated recommendation for this envelope.
    case suggested
    /// Zeroes the envelope out without having to subtract its balance by hand.
    case empty
}

/// The keypad's state machine, kept out of the view so it can be tested.
///
/// Amounts are entered the way a cash register takes them: digits fill in from
/// the right, so `1`, `2`, `3` reads as $1.23. Pressing `+` or `-` parks the
/// current value and starts a fresh entry; `=` resolves it, which is what lets
/// someone add to an existing envelope balance without doing the mental math.
public struct KeypadEngine: Hashable, Sendable {
    /// The amount currently shown on the big display.
    public private(set) var value: Money
    /// Set once the user starts typing over the incoming amount, so the first
    /// digit replaces the balance rather than appending to it.
    public private(set) var isTyping: Bool
    /// The parked left-hand side of a pending `+` / `-`.
    public private(set) var pendingValue: Money?
    public private(set) var pendingOperation: Operation?

    public enum Operation: Hashable, Sendable {
        case add
        case subtract
    }

    /// Largest amount the keypad accepts, guarding against runaway digit entry.
    public static let maximumMinorUnits = 99_999_999_99

    public init(value: Money = .zero) {
        self.value = value
        self.isTyping = false
        self.pendingValue = nil
        self.pendingOperation = nil
    }

    /// True while a `+` or `-` is waiting on `=`, which the view uses to show the
    /// running expression above the display.
    public var hasPendingOperation: Bool { pendingOperation != nil }

    public mutating func press(_ key: KeypadKey, suggested: Money = .zero) {
        switch key {
        case .digit(let digit):
            appendDigits(String(max(0, min(9, digit))))
        case .doubleZero:
            appendDigits("00")
        case .delete:
            deleteLastDigit()
        case .add:
            beginOperation(.add)
        case .subtract:
            beginOperation(.subtract)
        case .equals:
            resolvePendingOperation()
        case .suggested:
            replace(with: suggested)
        case .empty:
            replace(with: .zero)
        }
    }

    /// Discards any half-finished expression and returns the final amount, which
    /// is what `next` commits to the envelope.
    public mutating func commit() -> Money {
        resolvePendingOperation()
        isTyping = false
        return value
    }

    /// Loads the next envelope's balance in without carrying state across.
    public mutating func reset(to newValue: Money) {
        self = KeypadEngine(value: newValue)
    }

    // MARK: - Key handling

    private mutating func appendDigits(_ digits: String) {
        let existing = isTyping ? String(abs(value.minorUnits)) : "0"
        let combined = (existing == "0" ? "" : existing) + digits
        let trimmed = String(combined.prefix(11))
        let parsed = Int(trimmed) ?? 0
        value = Money(minorUnits: min(parsed, Self.maximumMinorUnits))
        isTyping = true
    }

    private mutating func deleteLastDigit() {
        guard isTyping else {
            // Nothing typed yet: the first delete clears the incoming balance,
            // matching how the real keypad behaves on a fresh envelope.
            value = .zero
            isTyping = true
            return
        }
        let digits = String(abs(value.minorUnits))
        let remaining = String(digits.dropLast())
        value = Money(minorUnits: Int(remaining) ?? 0)
    }

    private mutating func beginOperation(_ operation: Operation) {
        // Chaining without pressing equals resolves what came before first.
        resolvePendingOperation()
        pendingValue = value
        pendingOperation = operation
        value = .zero
        isTyping = false
    }

    private mutating func resolvePendingOperation() {
        guard let operation = pendingOperation, let left = pendingValue else { return }
        switch operation {
        case .add:
            value = left + value
        case .subtract:
            value = left - value
        }
        pendingValue = nil
        pendingOperation = nil
        isTyping = false
    }

    private mutating func replace(with newValue: Money) {
        pendingValue = nil
        pendingOperation = nil
        value = newValue
        isTyping = false
    }
}

File: web/src/domain/keypadEngine.ts
import type { Money } from './money'

/** The keys on the budgeting keypad. */
export type KeypadKey =
  | { type: 'digit'; digit: number }
  | { type: 'doubleZero' }
  /** Backspace — removes the last typed digit, or clears the amount if none have been typed yet. */
  | { type: 'delete' }
  | { type: 'add' }
  | { type: 'subtract' }
  | { type: 'equals' }
  /** Fills in the engine-calculated recommendation for this envelope. */
  | { type: 'suggested' }
  /** Zeroes the envelope out without having to subtract its balance by hand. */
  | { type: 'empty' }

export type KeypadOperation = 'add' | 'subtract'

/**
 * The keypad's state, kept out of the view so it can be tested.
 *
 * Amounts are entered the way a cash register takes them: digits fill in from
 * the right, so `1`, `2`, `3` reads as $1.23. Pressing `+` or `-` parks the
 * current value and starts a fresh entry; `=` resolves it, which is what lets
 * someone add to an existing envelope balance without doing the mental math.
 */
export interface KeypadState {
  /** The amount currently shown on the big display. */
  value: Money
  /**
   * Set once the user starts typing over the incoming amount, so the first
   * digit replaces the balance rather than appending to it.
   */
  isTyping: boolean
  /** The parked left-hand side of a pending `+` / `-`. */
  pendingValue: Money | null
  pendingOperation: KeypadOperation | null
}

/** Largest amount the keypad accepts, guarding against runaway digit entry. */
export const MAXIMUM_MINOR_UNITS = 99_999_999_99

export function keypadState(value: Money = 0): KeypadState {
  return { value, isTyping: false, pendingValue: null, pendingOperation: null }
}

/** True while a `+` or `-` is waiting on `=`. */
export function hasPendingOperation(state: KeypadState): boolean {
  return state.pendingOperation !== null
}

export function press(state: KeypadState, key: KeypadKey, suggested: Money = 0): KeypadState {
  switch (key.type) {
    case 'digit':
      return appendDigits(state, String(Math.max(0, Math.min(9, key.digit))))
    case 'doubleZero':
      return appendDigits(state, '00')
    case 'delete':
      return deleteLastDigit(state)
    case 'add':
      return beginOperation(state, 'add')
    case 'subtract':
      return beginOperation(state, 'subtract')
    case 'equals':
      return resolvePendingOperation(state)
    case 'suggested':
      return replaceValue(suggested)
    case 'empty':
      return replaceValue(0)
  }
}

/**
 * Resolves any half-finished expression and returns the final amount, which is
 * what `next` commits to the envelope.
 */
export function commit(state: KeypadState): { state: KeypadState; value: Money } {
  const resolved = resolvePendingOperation(state)
  return { state: { ...resolved, isTyping: false }, value: resolved.value }
}

/** Loads the next envelope's balance in without carrying state across. */
export function reset(value: Money): KeypadState {
  return keypadState(value)
}

function appendDigits(state: KeypadState, digits: string): KeypadState {
  const existing = state.isTyping ? String(Math.abs(state.value)) : '0'
  const combined = (existing === '0' ? '' : existing) + digits
  const trimmed = combined.slice(0, 11)
  const parsed = Number.parseInt(trimmed, 10) || 0
  return { ...state, value: Math.min(parsed, MAXIMUM_MINOR_UNITS), isTyping: true }
}

function deleteLastDigit(state: KeypadState): KeypadState {
  if (!state.isTyping) {
    // Nothing typed yet: the first delete clears the incoming balance,
    // matching how the real keypad behaves on a fresh envelope.
    return { ...state, value: 0, isTyping: true }
  }
  const digits = String(Math.abs(state.value))
  const remaining = digits.slice(0, -1)
  return { ...state, value: Number.parseInt(remaining, 10) || 0 }
}

function beginOperation(state: KeypadState, operation: KeypadOperation): KeypadState {
  // Chaining without pressing equals resolves what came before first.
  const resolved = resolvePendingOperation(state)
  return { value: 0, isTyping: false, pendingValue: resolved.value, pendingOperation: operation }
}

function resolvePendingOperation(state: KeypadState): KeypadState {
  if (state.pendingOperation === null || state.pendingValue === null) return state
  const value =
    state.pendingOperation === 'add'
      ? state.pendingValue + state.value
      : state.pendingValue - state.value
  return { value, isTyping: false, pendingValue: null, pendingOperation: null }
}

function replaceValue(value: Money): KeypadState {
  return { value, isTyping: false, pendingValue: null, pendingOperation: null }
}
